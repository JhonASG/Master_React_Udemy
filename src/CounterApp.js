import React, {useState} from "react";
import PropsTypes from "prop-types";

//Functional component
const CounterApp = ({value}) =>{
    const titulo = 'CounterApp';
    const [counter, setCounter] = useState(value);
    
    const sumador = () => {
        setCounter((counter)=>counter+1);
    };

    const restador = () => {
        setCounter((counter)=>counter-1);
    };

    const reset = () =>{
        setCounter(value);
    };

    return (
        <>
            <h1>{titulo}</h1>
            <h2>{counter}</h2>
            <button className="FirstButton" onClick={sumador}>More</button>
            <button className="FirstButton" onClick={restador}>Less</button>
            <button className="FirstButton" onClick={reset}>Reset</button>
        </>
    );
};

CounterApp.propTypes = {
    value: PropsTypes.number.isRequired
}

export default CounterApp;