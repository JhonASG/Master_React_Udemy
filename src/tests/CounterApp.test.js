import React from 'react';
import { render, screen, fireEvent } from "@testing-library/react";
import CounterApp from "../CounterApp";

describe('Test Homework', () => {
    let val = 10;

    let renderComponent = () => {
        return render(<CounterApp />);
    };

    // Este metodo se va a ejecutar antes de cada evaluaciÃ³n o test suite
    beforeEach(() => {
        renderComponent(); 
    });

    test('Muestra el valor por defecto del CounterAPP', () => {
        expect(screen.getByText(val)).toMatchSnapshot();
    });

    test('Muestra un valor de 100 y se verifica que sea el mostrado en el CounterAPP', () => {
        val = 100;

        render( <CounterApp value={val}/> );

        expect(screen.getByText(val)).toContainHTML('h2');
    });

    test('Debe incrementar con el boton more', () => {
        val = 11;
        const button = screen.getByText('More');

        fireEvent.click(button);

        expect(screen.getByText(val)).toContainHTML('h2');
    });

    test('Debe decrementar con el boton less', () => {
        val = 9;
        const button = screen.getByText('Less');

        fireEvent.click(button);

        expect(screen.getByText(val)).toContainHTML('h2');
    });

    test('Debe Resetear el valor por defecto con el boton reset', () => {
        val = 10;

        // Sumo uno al contador queda en 11 ya que el valor por defecto es 10
        let button = screen.getByText('More');
        fireEvent.click(button);

        // Con screen.getAllByRole('heading', {level: 2}) se obtienen todos los elementos h2 del JSX del componente
        // console.log(screen.getAllByRole('heading', {level: 2}));

        // Se resetea el contador pasando de 11 a su valor predeterminado que es 10
        button = screen.getByText('Reset');
        fireEvent.click(button);

        // Se evalua que el contenido en h2 que es donde se almacena el contador sea igual a 10 que es el valor por defecto
        expect(screen.getByText(val)).toContainHTML('h2');
    });
});