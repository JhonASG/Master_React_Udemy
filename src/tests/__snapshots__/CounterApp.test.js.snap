// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test Homework Muestra el valor por defecto del CounterAPP 1`] = `
<h2>
  10
</h2>
`;
